lib/testutils/src/test_backend.rs

```diff
@@ -249,6 +254,51 @@ impl Backend for TestBackend {
         Ok(id)
     }
 
+    async fn read_copy(&self, id: &CopyId) -> BackendResult<CopyHistory> {
+        let copy = self.locked_data().copies.get(id).cloned().ok_or_else(|| {
+            BackendError::ObjectNotFound {
+                object_type: "copy".to_string(),
+                hash: id.hex(),
+                source: "".into(),
+            }
+        })?;
+        Ok(copy)
+    }
+
+    async fn write_copy(&self, contents: &CopyHistory) -> BackendResult<CopyId> {
+        let id = CopyId::new(get_hash(contents));
+        self.locked_data()
+            .copies
+            .insert(id.clone(), contents.clone());
+        Ok(id)
+    }
+
+    async fn get_related_copies(&self, copy_id: &CopyId) -> BackendResult<Vec<CopyHistory>> {
+        let copies = &self.locked_data().copies;
+        if !copies.contains_key(copy_id) {
+            return Err(BackendError::ObjectNotFound {
+                object_type: "copy history".to_string(),
+                hash: copy_id.hex(),
+                source: "".into(),
+            });
+        }
+        // Return all copy histories to test that the caller correctly ignores histories
+        // that are not relevant to the trees they're working with.
+        let mut work: VecDeque<_> = copies.iter().collect();
+        let mut histories = vec![];
+        let mut emitted = HashSet::new();
+        while let Some((id, copy)) = work.pop_front() {
+            if copy.parents.iter().all(|id| emitted.contains(id)) {
+                histories.push(copy.clone());
+                emitted.insert(id);
+            } else {
+                work.push_back((id, copy));
+            }
+        }
+        histories.reverse();
```

Comment:
nit: maybe better to use `topo_order_reverse()` to sort ids (or id/object pairs) first? Quadratic sorting wouldn't matter in tests, though.

Reply:
It's simpler anyway, so I've switched to that function.